### ISSUE TITLE
Provide a way to cancel NotifyOnStateChange

### ISSUE THREAD

**Issue by noahdav** [grpc/grpc#21948]
### What version of gRPC and what language are you using?
1.27.0-dev commit b94490bc74a2efc7274af70f276eb4c14c226d6f

### What operating system (Linux, Windows,...) and version?
Windows

### What runtime / compiler are you using (e.g. python version or version of gcc)
MSVC 2017

### What did you do?
I created a completion queue to monitor the channel state.  I registered the channel state change on the completion queue with a max timeout.  I want to know whenever the state changes so that I can notify any listening clients.  I am trying to wait indefinitely so that I do not miss any state changes due to timeouts.

I registered by
channel->NotifyOnStateChange(channel->GetState(true), std::chrono::system_clock::time_point::max(), &m_cq, ptr.get());

During my application shutdown, I am calling m_cq.Shutdown().  At this point I expect that CompletionQueue::Next should false since it has been shutdown.  

### What did you expect to see?
I expected that the thread waiting on CompletionQueue::Next should be unblocked since the completion queue is shutdown.  As I understand this from the documentation false should be returned.

### What did you see instead?

The call to CompletionQueue::Next is stuck waiting for a state change.  If I force a state change by killing the server it will return however I need to be able to shutdown gracefully even when no state change occurs.

**Comment by noahdav** [comment]
In my case I have a client which wants to shut down but there is a destructor waiting to join on the tread which is monitoring the completion queue.  Since the completion queue does not get the shutdown message the tread will not exit.  

I tried to debug this more and if the NotifyOnStateChange is not in the completion queue it shutdowns cleanly

```
        grpc::CompletionQueue cq;
        std::promise<bool> start;
        std::promise<bool> complete;

        std::thread t([&cq, &start, &complete]() {
            void* tag = nullptr;
            bool ok = false;

            start.set_value(true);
            while (cq.Next(&tag, &ok))
            {
                continue;
            }
            complete.set_value_at_thread_exit(true);
        });
        
        // wait for thread to start
        start.get_future().wait();
        cq.Shutdown();

        // wait for exit after we shutdown q
        complete.get_future().wait();
        t.join();
```

However if the NotifyOnStateChange is in the completion queue then it will block waiting for the state change notification.

```
        rpc::CompletionQueue cq;
        std::shared_ptr<grpc::Channel> channel = grpc::CreateChannel("localhost:50001", grpc::InsecureChannelCredentials());
        channel->NotifyOnStateChange(channel->GetState(true), std::chrono::system_clock::time_point::max(), &cq, nullptr);

        std::promise<bool> complete;
        std::promise<bool> state_change;

        std::thread t([&cq, channel, &state_change, &complete]() {
            void* tag = nullptr;
            bool ok = false;

            while (cq.Next(&tag, &ok))
            {
                grpc_connectivity_state state = channel->GetState(true);
                channel->NotifyOnStateChange(state, std::chrono::system_clock::time_point::max(), &cq, nullptr);
                if(state == GRPC_CHANNEL_READY)
                    state_change.set_value(true);
            }
            complete.set_value_at_thread_exit(true);
        });

        // wait for channel to be in ready state before we shutdown q
        state_change.get_future().wait();

        cq.Shutdown();

        // wait for exit after we shutdown q
        complete.get_future().wait();
        t.join();
```

**Comment by mehrdada** [comment]
I suppose there's a race between `cq.Shutdown` in main thread in your code and passing the CQ in your `cq.Next` thread when it calls `NotifyOnStateChange`.  Once you start calling `Shutdown` on your CQ, you're not supposed to use it for other operations.

**Comment by mehrdada** [comment]
Scratch my last message -- not sure if that's actually the case.

**Comment by vjpai** [comment]
Hi there,

Mehrdad's response above was partially correct. Shutdown means that you're not supposed to use the CQ for new operations. However, it does _not _ short-circuit or fail existing operations. You have to shutdown the CQ and then drain it, and the draining will require the completion of existing operations. In the case of alarms and RPCs, you can force their prompt completion by canceling them. This case looks like a bit of a problem since we don't have a mechanism for canceling them. So, basically, I'm saying that this works according to spec but maybe the spec is not what it should be. I'm going to reclassify this as a feature request and change the title.

**Comment by noahdav** [comment]
@vjpai I suppose that I could use AsyncNext with a timeout and check if I want to shutdown.  However this will not drain the queue but simply short circuit the check.  What would be the side effects of doing this as a workaround?

**Comment by vjpai** [comment]
That would likely cause a memory leak. I'd instead suggest using a shorter timeout than timepoint::max() if that's a viable option.

**Comment by noahdav** [comment]
Thank you @vjpai .  I changed this to use the shorter timeout approach in the meantime.  I am not sure from the documentation so I just wanted to verify that cq.Next() will only return false after shutdown.  Is there any other reasons why Next() will return false?

**Comment by vjpai** [comment]
Yes, next will only return false when the completion queue is shut down and fully drained.

**Comment by mehrdada** [comment]
@vjpai I think this can fit into the broader issue that's super essential when we need a clean shutdown process: `Channel::Close()` which terminates all active calls and ops like this. https://github.com/grpc/grpc/issues/21926

Right now we'll have to have an additional layer of tracking all the active calls unnecessarily introducing additional complexity and perf burden when core can simply disconnect the underlying TCP connection and have every pending op fail.

**Comment by mehrdada** [comment]
I suppose this is now a duplicate of age-old issue: https://github.com/grpc/grpc/issues/3064 -- I'll close in favor of the original 4-digit issue.
